docs/terminals.md

```diff
@@ -0,0 +1,179 @@
+
+For example:
+
+```
+$ runc create foo              # container created, inheriting stdio from this terminal
+$ runc start foo               # container started, inheriting stdio from this terminal
+$ curl https://containerd.io   # get contents of this site to the terminal
+
+<!DOCTYPE html>
+I am container output
+<html lang="en">
+  <head>
+	  ASashcned more container output
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>containerd</title>
+```
+
+Two problems occurred above:
+
+1. The output from the `curl` command and the output of the container `foo` became intermingled.
+2. Our typing `curl https://containerd.io` actually was passed to both our shell and the container `foo`. The results can be unpredictable. Our example here was simulated and might not even work.
+
+Further, a third, possibly temporary problem exists. Pass-Through Detached only works correctly when stdout/stderr are either an actual terminal or a file. If you use _anything_ else, e.g. a pipe, it will hang, for reasons unknown. There is an [open issue](https://github.com/opencontainers/runc/issues/1721) on this.
+
+For this reason, we **strongly recommend** following one of these when running `runc` detached:
+
+* Use New Terminal mode.
+* Set up your own separate dedicated stdio and pass it through.
+
+#### New Terminal Detached
+When running runc detached with New Terminal mode, runc behaves exactly as described above, creating a new pty, passing it to the container, sending the fd to a socket, and exiting.
+
+## How To Get the FD for New Terminal
+When running `runc` in New Terminal mode, i.e. `terminal:true` in `config.json` - whether foreground or detached `runc` - you **must** tell `runc` what process will control the master end of the pty. To do that:
+
+1. Open a Unix-domain socket
+2. Pass the path to the socket to `runc create <container> --console-socket <path_to_socket>`
+3. Listen on the socket to get the master file descriptor
+4. Use that file descriptor to control stdio
+
+Note the one shortcoming: the single descriptor means that you have mixed stdout/stderr. There is no way to separate them. Essentially, this is a _console_, which has no separate stdout and stderr. This is the same issue you have with any console, or even any normal terminal window. To see the issue:
+
+```
+/ # ls -l /proc/self/fd/[12]
+lrwx------    1 root     root            64 Mar 19 07:09 /proc/self/fd/1 -> /0
+lrwx------    1 root     root            64 Mar 19 07:09 /proc/self/fd/2 -> /0
+```
+
+The following sample code, partially taken from the reference implementation [recvtty](https://github.com/opencontainers/runc/blob/master/contrib/cmd/recvtty/recvtty.go), shows how to accept a file descriptor for the socket. It does not contain the usual `err` handling and `defer`red closing.
```

Comment:
i would prefer to just point directly to `recvtty.go` rather than attaching sample code here so there are one less place to keep change.